inc/Modules/Rest/Abstract_REST_Controller.php

```diff
@@ -0,0 +1,136 @@
+<?php
+/**
+ * Base REST controller class.
+ *
+ * Includes the shared namespace, version and hook registration.
+ *
+ * @package OneMedia\Modules\Rest
+ */
+
+declare( strict_types = 1 );
+
+namespace OneMedia\Modules\Rest;
+
+use OneMedia\Contracts\Interfaces\Registrable;
+use OneMedia\Modules\Settings\Settings;
+
+/**
+ * Class - Abstract_REST_Controller
+ */
+abstract class Abstract_REST_Controller extends \WP_REST_Controller implements Registrable {
+	/**
+	 * The namespace for the REST API.
+	 */
+	public const NAMESPACE = 'onemedia/v1';
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function register_hooks(): void {
+		add_action( 'rest_api_init', [ $this, 'register_routes' ] );
+	}
+
+	/**
+	 * {@inheritDoc}
+	 *
+	 * We throw an exception here to force the child class to implement this method.
+	 *
+	 * @throws \Exception If method not implemented.
+	 *
+	 * @codeCoverageIgnore
+	 */
+	public function register_routes(): void {
+		throw new \Exception( __FUNCTION__ . ' Method not implemented.' );
+	}
+
+	/**
+	 * Checks for the use of the OneMedia API key in the request headers.
+	 *
+	 * @todo this should be on a hook.
+	 *
+	 * @param \WP_REST_Request<array{}> $request Request.
+	 * @return bool
+	 */
+	public function check_api_permissions( $request ) {
```

Comment:
Backport the latest version of this from OneSearch/OneUpdate that handles "true" cross-site. then make sure that any server-side requests ( i.e. `wp_{safe}_remote_*() calls) have an explicit `headers['origin'] => get_site_url()` set (see those plugins examples).